src/include/hardware.h

```diff
@@ -31,19 +31,40 @@
  * same memory location, but to do this safely we need to precisely control how reads to memory are
  * performed. This macro gives us control over this.
  *
- * GCC and Clang have a __typeof__ compiler builtin which allows us to temporarily cast the value
- * being read as a volatile and achieve volatile semantics. For other compilers we'll fall back on
- * inserting a read barrier after the read (our pre-existing implementation) which prevents invented
- * and fused loads for this variable in the code following the expression.
+ * To allow this macro to be used in a backwards compatible way it has been implemented using
+ * barriers. Technically this can be achieved by a volatile cast.
+ */
+#define WT_READ_ONCE(v, val) \
+    do {                     \
+        WT_READ_BARRIER();   \
+        (v) = (val);         \
+        WT_READ_BARRIER();   \
+    } while (0)
+
+/*
+ * WT_WRITE_ONCE --
+ *
+ * Ensure a single write to memory in the source code produces a single write to memory in the
+ * compiled output.
  *
- * FIXME-WT-11718 - Once Windows build machines that support C11 _Generics are available this macro
- * will be updated to use _Generic on all platforms.
+ * We can describe a scenario where a variable in memory is read to a local register, and then the
+ * register value is repeatedly updated before the final value is written back to memory. Without a
+ * WT_WRITE_ONCE wrapping the final write back to memory, the compiler is allowed to convert these
+ * register writes into writes to memory (see "register spilling") which can leak interim, invalid,
```

Comment:
My reading about "register spilling" (e.g: https://en.wikipedia.org/wiki/Register_allocation#:~:text=If%20there%20are%20not%20enough,then%20considered%20as%20%22split%22.) indicates that the content of the register is spilled to an allocated of memory used for managing code that requires more registers than the hardware has available.
Not that the content gets "spilled" back to a shared memory location.
Is there a different version of spilling? If so could you point me to a description?

Reply:
This is largely borrowed from the [perf books](https://mirrors.edge.kernel.org/pub/linux/kernel/people/paulmck/perfbook/perfbook.2023.06.11a.pdf) section on invented stores which is under section 4.3.4, where it gives an example of the compiler borrowing a shared memory location as nothing reads from that location in the given context. Its quite likely that this example wouldn't happen in reality.